### PR TITLE
Update to use other forms of calls, such as nodejs, as a native module

### PR DESCRIPTION
**Pull request by it-fm**
I have made some changes so that other forms of calling and exporting are compatible, I am new to c ++, I need feedback if the proposed change is valid and correct.

### CODE REVIEW

File: lib/picture.cpp [hunk 1]
```diff
@@ -63,6 +63,37 @@ void Picture::save(const std::string& filename)
     fclose(fd);
 }
 
+std::vector<uint8_t> Picture::exportEncoded()
```
Comment (krepa098) [inline lib/picture.cpp]: Buffer Picture::exportEncoded()

File: lib/picture.cpp [hunk 2]
```diff
@@ -63,6 +63,37 @@ void Picture::save(const std::string& filename)
...
+    auto info_ptr = png_create_info_struct(png_ptr);
+
+    std::vector<uint8_t> encoded;
+    // This callback will be called each time libpng wants to write an encoded chunk.
+    // https://github.com/Prior99/node-libpng/blob/master/native/encode.cpp
+    png_set_write_fn(png_ptr, &encoded, [] (png_structp png_ptr, png_bytep data, png_size_t length) {
+        auto encoded = reinterpret_cast<std::vector<uint8_t>*>(png_get_io_ptr(png_ptr));
+        encoded->insert(encoded->end(), data, data + length);
+    }, nullptr);
+
```
Comment (krepa098) [inline lib/picture.cpp:78]: Make sure to run clang-format on the files you change (if your IDE doesn't do it automatically). 
e.g. `clang-format -i ./lib/picture.cpp` it should pick up the style automatically

File: lib/picture.cpp [hunk 3]
```diff
@@ -63,6 +63,37 @@ void Picture::save(const std::string& filename)
     fclose(fd);
 }
 
+std::vector<uint8_t> Picture::exportEncoded()
+{
+    auto png_ptr  = png_create_write_struct(PNG_LIBPNG_VER_STRING, nullptr, nullptr, nullptr);
+    auto info_ptr = png_create_info_struct(png_ptr);
+
+    std::vector<uint8_t> encoded;
```
Comment (krepa098) [inline lib/picture.cpp]: Buffer encoded;

File: lib/picture.cpp [hunk 4]
```diff
@@ -63,6 +63,37 @@ void Picture::save(const std::string& filename)
...
+    // This callback will be called each time libpng wants to write an encoded chunk.
+    // https://github.com/Prior99/node-libpng/blob/master/native/encode.cpp
+    png_set_write_fn(png_ptr, &encoded, [] (png_structp png_ptr, png_bytep data, png_size_t length) {
+        auto encoded = reinterpret_cast<std::vector<uint8_t>*>(png_get_io_ptr(png_ptr));
```
Comment (krepa098) [inline lib/picture.cpp]: Buffer*

File: lib/picture.h [hunk 5]
```diff
@@ -39,6 +39,7 @@ class Picture
 
     Byte* data();
     void save(const std::string& filename);
+    std::vector<uint8_t> exportEncoded();
```
Comment (krepa098) [inline lib/picture.h]: Buffer exportEncoded();

File: lib/stl/parser.cpp [hunk 6]
```diff
@@ -43,7 +43,15 @@ Mesh Parser::parseFile(const std::string& filename) const
     return parseAscii(stream);
 }
 
-bool Parser::isBinaryFormat(std::ifstream& in) const
+Mesh Parser::parseStream(std::istream& in) const
```
Comment (krepa098) [inline lib/stl/parser.cpp:41]: `parseFile()` should make use of `parseStream()`

**Review by krepa098 — APPROVED** [review]
Thank you. Now it's clear what you're trying to achieve. I'll merge this once the changes are addressed.

File: lib/picture.cpp (hunk 4 above)
Comment (it-fm) [reply lib/picture.cpp]: I have the following error: `no suitable constructor exists to convert from "std::vector<Byte, std::allocator<Byte>> *" to "std::vector<Byte, std::allocator<Byte>>"`


however when creating a pointer I am successful in compiling and working: `Buffer * encoded`
Any suggestion ?

Comment (krepa098) [reply lib/picture.cpp]: `auto encoded = reinterpret_cast<Buffer*>(png_get_io_ptr(png_ptr));` works fine I suppose?

File: lib/stl/parser.cpp (hunk 6 above)
Comment (krepa098) [reply lib/stl/parser.cpp:41]: Thanks. Don't forget to run clang-format on that file as well.

**Review by it-fm — COMMENTED** [review]
ready, make sure everything is okay

File: lib/stl/parser.cpp (hunk 6 above)
Comment (krepa098) [reply lib/stl/parser.cpp:41]: This file still needs clang-format (see the white spaces e.g.)

**Review by it-fm — COMMENTED** [review]
I put .clang-format in the root folder and run `clang-format -i stl / parser.cpp`

**Comment by krepa098** [comment]
Interesting to see that we don't get the same result from clang-format. Well, thanks for trying, I'll merge it now.

**Pull request merged**